lib/hanami/slice.rb

```diff
@@ -1,148 +1,201 @@
 # frozen_string_literal: true
 
 require "dry/system/container"
+require "hanami/errors"
 require "pathname"
+require_relative "constants"
 
 module Hanami
   # Distinct area of concern within an Hanami application
   #
   # @since 2.0.0
   class Slice
-    attr_reader :application, :name, :namespace, :root
-
-    def initialize(application, name:, namespace: nil, root: nil, container: nil)
-      @application = application
-      @name = name.to_sym
-      @namespace = namespace
-      @root = root ? Pathname(root) : root
-      @container = container
-    end
+    class << self
+      attr_reader :container
 
-    def inflector
-      application.inflector
-    end
+      def inherited(klass)
+        super
 
-    def namespace_path
-      @namespace_path ||= inflector.underscore(namespace.to_s)
-    end
+        # Create a (to be configured later) container as early as possible, since code in
+        # the slice subclass may want to start referring to it right away (e.g. `.import`)
+        klass.instance_variable_set(:@container, Class.new(Dry::System::Container))
+      end
 
-    def prepare(provider_name = nil)
-      if provider_name
-        container.prepare(provider_name)
-        return self
+      def application
```

Comment:
@timriley Shouldn't this be injected as dependency?

Reply:
@jodosha Ah yes, thanks for raising this. I hard-coded it while doing all this refactoring, as a matter of expediency, but it does make sense to allow this to be configured.

Since this is class-level state in the new slice structure, we can't "inject" the application in the typical sense, but I've have made it a class-level ivar and changed how we pass it through in https://github.com/hanami/hanami/pull/1150/commits/7eade29833fc2db90d579732457a462a98bde138:

1. It's set as a class-level ivar in the `.inherited` hook like the container
2. It can be explicitly reassigned via `.application(new_application)` if needed (this capacity will not be user-visible for now, but if we ever reach a future where we want to support multiple distinct Hanami applications in memory, this is how we can do it for slices)
3. The slice registrar uses the above method to assign the application when building slice classes dynamically
4. And finally, the slice registrar will raise an exception if it ever receives a slice that has a _different_ application assigned

Together I feel this makes for a pretty robust approach. Let me know what you think :)

Reply:
For the record, I discussed this with @jodosha and this original question was based on a misunderstanding. I've just pushed https://github.com/hanami/hanami/pull/1150/commits/a8d67df0db441cfa73d8e8e498f630e0bdc76a8a which goes back to slices all having a fixed reference to the single `Hanami.application`.

This does keep a couple of structural changes from the first commit that I liked:

- Keeping most of the `Slice` behaviour in a `ClassMethods` module that is extended with `Slice` is inherited.
- Providing access to the application by assigning it to a class-level ivar when Slice is inherited (rather than having a method with a hard-coded reference). This at least makes it clearer that the `@application` is some externally provided state, which is a useful signal should we decide we want to do anything different with it in future.